core/atomic_c11.c

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright (C) 2015-2016 Eistec AB
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup         core_util
+ * @{
+ *
+ * @file
+ *
+ * @brief Implementation of C11 atomic operations if GCC does not provide an
+ * implementation.
+ *
+ * GCC with -mcpu=cortex-m3 and cortex-m4 generate LDREX/STREX instructions
+ * instead of library calls. There is however currently (2015-05-29) no
+ * implementation for Cortex-M0, since it lacks the lock-free atomic operations
+ * found in the M3 and M4 cores.
+ *
+ * @note Other CPUs (e.g. msp430, avr) might need this too, but current MSP430
+ * GCC in Ubuntu/Debian is stuck at version 4.6 which does not provide C11
+ * language support which makes it difficult to actually make use of this on
+ * that platform.
+ *
+ * @note This implementation completely ignores the memory model parameter
+ *
+ * @see https://gcc.gnu.org/wiki/Atomic/GCCMM/LIbrary
+ * @see https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html
+ *
+ * @author Joakim Nohlgård <joakim.nohlgard@eistec.se>
+ */
+
+#include <stddef.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdatomic.h>
+#include <string.h>
+#include "irq.h"
+
+/* GCC documentation refers to the types as I1, I2, I4, I8, I16 */
+typedef uint8_t  I1;
+typedef uint16_t I2;
+typedef uint32_t I4;
+typedef uint64_t I8;
+/* typedef __uint128_t I16; */ /* No 128 bit integer support yet */
+
+/**
+ * @brief This is a macro that defines a function named __atomic_load_<em>n</em>
+ *
+ * @param[in]  n         width of the data, in bytes
+ */
+#define TEMPLATE_ATOMIC_LOAD_N(n) \
```

Comment:
I agree that macros make sense here, but just for the record: I hate macros and if they will ever create any problems - I told you so! ;-)